cache/ttl/ttl.go

```diff
@@ -0,0 +1,162 @@
+package ttl
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/renproject/kv/db"
+)
+
+var (
+	// PrunePointerTableName is the name of the table where we store the prune pointer.
+	PrunePointerTableName = "ttl_0"
+
+	// PrunePointerKey is the key of the key-value pair which we can use to
+	// query the current prune pointer.
+	PrunePointerKey = "prune_pointer"
+)
+
+type inMemTTL struct {
+	timeToLive    time.Duration
+	pruneInterval time.Duration
+	codec         db.Codec
+	db            db.DB
+}
+
+// New returns a new ttl wrapper over the given database.
+// The underlying database cannot have any database has a prefix of `ttl_`.
+func New(ctx context.Context, database db.DB, timeToLive time.Duration, pruneInterval time.Duration, codec db.Codec) db.DB {
+	ttlDB := inMemTTL{
+		timeToLive:    timeToLive,
+		pruneInterval: pruneInterval,
+		codec:         codec,
+		db:            database,
+	}
+
+	ttlDB.prune(ctx)
+
+	return &ttlDB
+}
+
+func (ttlDB *inMemTTL) Table(name string) db.Table {
+	return ttlDB.db.Table(name)
+}
+
+// Insert implements the `db.table` interface.
+func (ttlDB *inMemTTL) Insert(name string, key string, value interface{}) error {
+
+	// Insert the new data entry into the underlying DB.
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+	if err := ttlDB.db.Insert(name, key, value); err != nil {
+		return err
+	}
+
+	// Insert the current timestamp for future pruning.
+	slot := ttlDB.slotNo(time.Now())
+	slotTableName := fmt.Sprintf("ttl_%d", slot)
+	return ttlDB.db.Insert(slotTableName, key, name)
+}
+
+// Get implements the `db.table` interface.
+func (ttlDB *inMemTTL) Get(name string, key string, value interface{}) error {
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+
+	return ttlDB.db.Get(name, key, value)
+}
+
+// Delete implements the `db.table` interface.
+func (ttlDB *inMemTTL) Delete(name string, key string) error {
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+
+	return ttlDB.db.Delete(name, key)
+}
+
+// Size implements the `db.table` interface.
+func (ttlDB *inMemTTL) Size(name string) (int, error) {
+	return ttlDB.db.Size(name)
+}
+
+// Iterator implements the `db.table` interface.
+func (ttlDB *inMemTTL) Iterator(name string) (db.Iterator, error) {
+	return ttlDB.db.Iterator(name)
+}
+
+// prune will periodically prune the underlying database and stores the prune pointer
+// in the db.
+func (ttlDB *inMemTTL) prune(ctx context.Context) {
+	pointer, err := ttlDB.prunePointer()
+	if err != nil {
+		log.Fatalf("cannot read prune pointer, err = %v", err)
+	}
+
+	ticker := time.NewTicker(ttlDB.pruneInterval)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				err := func() error {
+					newSlotToDelete := ttlDB.slotNo(time.Now().Add(-1 * ttlDB.pruneInterval))
+					for slot := pointer + 1; slot <= newSlotToDelete; slot++ {
```

Comment:
Is there a reason for `slot := pointer+1; slot <= newSlotToDelete` instead of `slot := pointer; slot < newSlotToDelete`? The latter is a bit more obvious and gives us a "grace interval" (which we need because slots are rounded down by the current time, so things in slot `N` may have actually been inserted later than `N`).

Reply:
The grace interval has been included in 
```golang
newSlotToDelete := ttlDB.slotNo(time.Now().Add(-1 * ttlDB.pruneInterval))
```
Everything in the `newSlotToDelete` should already passed the `ttl` and we won't add any data into it. So it's safe to delete everything till the `newSlotToDelete`(included) and marked it as the new `prunePointer`.